bin/propolis-server/src/main.rs

```diff
@@ -86,6 +86,16 @@ async fn main() -> anyhow::Result<()> {
                 |error| anyhow!("failed to create logger: {}", error),
             )?;
 
+            // Check that devices conform to expected API version
+            if let Err(e) = propolis::vmm::version::check() {
+                slog::error!(log, "Error checking VMM API version");
+                return Err(e.into());
+            }
+            if let Err(e) = propolis::hw::virtio::viona::version::check() {
```

Comment:
Perhaps merge these checks into a single method the main propolis lib exports? Having every consumer explicitly check for sub component versions (like viona) seems like we're overly leaking some implementation details.

Reply:
The versioning for those two components are independent, so the consumer may want to make differing decisions about what's required.  (I almost made the propolis-standalone version check to see if there was even a viona device in the config in order to trigger the check)

Reply:
Hmm, perhaps something like:

```rust
pub enum VersionError {
    Io(std::io::Error),
    VmmMismatch(u32, u32),
    ComponentMismatch(&'static str, u32, u32),
}

fn version_check() -> Result<(), VersionError> {
    vmm_version::check()?;
    viona_version::check()?;
    Ok(())
}
```

With appropriate from impls for vmm/viona version errors, e.g.:
```rust
impl From<viona_version::VersionError> for VersionError {
    fn from(err: viona_version::VersionError) -> Self {
        use viona_version::VersionError as VionaVerErr;
        match err {
            VionaVerErr::Io(e) => VersionError::Io(e),
            VionaVerErr::Mismatch(a, b) => VersionError::ComponentMismatch("viona", a, b),
        }
    }
}
```

So server can just fail on error from `version_check` but standalone can check for viona (either just the string or replace with another enum if you want) specific errors and ignore them if it wants.

Reply:
How's the reworked stuff look?